File: CPP/ByTier/C/1688_C_Manipulating_History.cpp
// Problem: CF 1688 C - Manipulating History
// https://codeforces.com/contest/1688/problem/C

/*
Code Purpose:
This code determines the initial string of Gensokyo by exploiting the fact that each operation replaces a substring with another, and the sequence of operations is shuffled. The key insight is that the XOR of all characters in the operation strings (t_i) will cancel out all characters except the initial character of the original string. This works because each character in the original string appears an equal number of times in the t_i strings (once as a substring, once in the replacement), except for the root character which only appears in t_1.

Algorithms/Techniques:
- XOR operation to isolate the initial character
- Observing that t_1 (the first operation's first string) must be a substring of the original string and the replacement of it must include the initial character

Time Complexity: O(total length of all input strings)
Space Complexity: O(1) - only using a single char variable for XOR operations

*/

#include <iostream>
#include <string>

using namespace std;
int main() {
  int t;
  cin >> t;
  while (t--) {
    int n;
    cin >> n;
    n = 2 * n + 1;  // Total number of strings in t (2*n operations + 1 final string)
    char c = 0;     // Initialize XOR accumulator
    while (n--) {
      string s;
      cin >> s;
      for (auto x : s)
        c = c ^ x;  // XOR each character to cancel out pairs
    }
    cout << c << '\n';  // The remaining character is the initial one
  }
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1688/1688_C_Manipulating_History.py
# Problem: CF 1688 C - Manipulating History
# https://codeforces.com/contest/1688/problem/C

"""
Code Purpose:
This code solves the problem of reconstructing the initial string from a sequence of operations performed on it. 
The key insight is that each operation consists of a substring to be replaced (t_{2i-1}) and the replacement string (t_{2i}).
Since the operations are shuffled, we need to find the original string by reversing the process.
However, a simpler observation is that in the set of all strings in t and s, characters that appear an odd number of times
must be the ones that were not replaced, i.e., they are part of the original string.
Time Complexity: O(sum of lengths of all input strings)
Space Complexity: O(1) - since we use a fixed-size array of 26 characters
"""

for __ in range(int(input())):  # Process each test case
    n = 2 * int(input()) + 1  # Total number of strings in t plus final string s (n operations = 2*n strings in t, plus final string)
    a = [0 for i in range(26)]  # Array to count frequency of each character (a-z)
    
    for i in range(n):  # Read all 2*n + 1 strings (t_i)
        s = input()
        for c in s:  # For each character in the string
            a[ord(c) - ord("a")] += 1  # Increment count for that character
    
    cnt = 0  # Counter to track how many characters appear an odd number of times
    for i in range(26):  # Check each character from 'a' to 'z'
        if a[i] % 2 == 1:  # If character count is odd
            print(chr(i + ord("a")))  # Print the character
            cnt += 1  # Increment counter
    
    if cnt != 1:  # If not exactly one character appeared an odd number of times, the input is inconsistent
        print("fake problem")


# https://github.com/VaHiX/CodeForces/